### ISSUE TITLE
`in` arguments crash the compiler when temp stores are required

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#29371]
When trying to build one of my solution's code libraries, csc.exe exits with errorlevel 0x‭80131623‬, and 'dotnet build shows these stack traces:

```
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error : FailFast: System.InvalidOperationException: Unexpected value '4' of type 'Microsoft.CodeAnalysis.RefKind' [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.SyntheticBoundNodeFactory.StoreToTemp(BoundExpression argument, BoundAssignmentOperator&amp; store, RefKind refKind, SynthesizedLocalKind kind, SyntaxNode syntaxOpt) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.BuildStoresToTemps(Boolean expanded, ImmutableArray`1 argsToParamsOpt, ImmutableArray`1 parameters, ImmutableArray`1 argumentRefKinds, ImmutableArray`1 rewrittenArguments, Boolean forceLambdaSpilling, BoundExpression[] arguments, ArrayBuilder`1 refKinds, ArrayBuilder`1 storesToTemps) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.MakeArguments(SyntaxNode syntax, ImmutableArray`1 rewrittenArguments, Symbol methodOrIndexer, MethodSymbol optionalParametersMethod, Boolean expanded, ImmutableArray`1 argsToParamsOpt, ImmutableArray`1&amp; argumentRefKindsOpt, ImmutableArray`1&amp; temps, Boolean invokedAsExtensionMethod, ThreeState enableCallerInfo) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitObjectCreationExpression(BoundObjectCreationExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundObjectCreationExpression.Accept(BoundTreeVisitor visitor) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32&amp; recursionDepth, BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionImpl(BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitObjectCreationExpression(BoundObjectCreationExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundObjectCreationExpression.Accept(BoundTreeVisitor visitor) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32&amp; recursionDepth, BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionImpl(BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitCall(BoundCall node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundCall.Accept(BoundTreeVisitor visitor) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
C:\Program Files\dotnet\sdk\2.1.400-preview-009171\Roslyn\Microsoft.CSharp.Core.targets(52,5): error :    at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32&amp; recursionDepth, BoundExpression node) [E:\projects\testframework-wow\src\Blizzard.TestFramework.WoW\Blizzard.TestFramework.WoW.csproj]
...
```

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/306848/unexpected-value-4-of-type-microsoftcodeanalysisre.html
VSTS ticketId: 659932_
_These are the original issue comments:_

Daniel Hobbs on 8/15/2018, 08:47 AM (2 days ago): I had a similar problem. (Unexpected value '4' of type 'Microsoft.CodeAnalysis.RefKind') I think it may be due to lowering with 'in'. I'm on VS 15.8. Here is a repro.

```C#
public readonly struct S
{        
    private static readonly S s_Default = default;        
    public static ref readonly S Default => ref s_Default;
}    
public class C
{
    public C(in S s, int i = 0)\{ }    
}    
public static class Foo 
{        
    // Crashes CSC - "csc.exe" exited with code -2146232797        
    public static C Instance { get; } = new C(s: in S.Default); // use 'in' and omit optional arg    
}
```

_These are the original issue solutions:_
(no solutions)

**Comment by jaredpar** [comment]
Can this be closed now?

**Comment by glenn-slayden** [comment]
Can I ask, what's the timeframe for this fix--and others from this repo for that matter--making their way into the mainstream vs2017 public release?

FWIW, my vs2017 install is currently reporting 15.8.7.

**Comment by jasonmalinowski** [comment]
This was merged into our branch for the Visual Studio "15.9" builds, which is currently in preview now. I can't quite tell if the fix got into Preview 3; it'll definitely be in Preview 4.

**Comment by dongyiqi** [comment]
happened when using Unity Editor 2019.3.8f1
